### PR TITLE
fix: tuple supported for getitem for Pandas dataframes

### LINKED ISSUE #990: [Bug]: Can not subset using `tuple()`
### Describe the bug

[Docs say](https://narwhals-dev.github.io/narwhals/api-reference/dataframe/#narwhals.dataframe.DataFrame.__getitem) you can subset using a `Sequence`. This should support `tuple()`, not just `list()`

### Steps or code to reproduce the bug

```python
import narwhals as nw
import pandas as pd
import polars as pl

DATA={ "x": [1,2,3] }
pd_df = nw.from_native(pd.DataFrame(DATA), eager_only=True)
pl_df = nw.from_native(pl.DataFrame(DATA), eager_only=True)

print("nw tuple subset")

print("list, list ; √")
print("tuple, list ; X")
print("list, tuple ; X")
print("tuple, tuple ; X")
print("")

tuple_row = (0,2)
list_row = list(tuple_row)
tuple_col = (0,)
list_col = list(tuple_col)

print("pd_df[list_row, list_col].shape:", pd_df[list_row, list_col].shape)
try:
    pd_df[tuple_row, list_col].shape
except pd.errors.IndexingError as e:
    print("pd_df[tuple_row, list_col].shape Error:", str(e))
try:
    pd_df[list_row, tuple_col].shape
except pd.errors.IndexingError as e:
    print("pd_df[list_row, tuple_col].shape Error:", str(e))
try:
    pd_df[tuple_row, tuple_col].shape
except pd.errors.IndexingError as e:
    print("pd_df[tuple_row, tuple_col].shape Error:", str(e))
```

### Expected results

All four examples should not throw an error and instead subset just like the list indexers

### Actual results

```
nw tuple subset
list, list ; √
tuple, list ; X
list, tuple ; X
tuple, tuple ; X

pd_df[list_row, list_col].shape: (2, 1)
pd_df[tuple_row, list_col].shape Error: Too many indexers
pd_df[list_row, tuple_col].shape Error: Too many indexers
pd_df[tuple_row, tuple_col].shape Error: Too many indexers
```

### Please run narwhals.show_version() and enter the output below.

```shell
System:
    python: 3.12.4 (main, Jun  6 2024, 18:26:44) [Clang 15.0.0 (clang-1500.3.9.4)]
executable: /path/to/python
   machine: macOS-14.6.1-arm64-arm-64bit

Python dependencies:
     narwhals: 1.8.1
       pandas: 2.2.2
       polars: 1.5.0
         cudf: 
        modin: 
      pyarrow: 
        numpy: 2.1.0
```
```


### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by mikeweltevrede**
## What type of PR is this? (check all applicable)

- [x] 💾 Refactor
- [ ] ✨ Feature
- [x] 🐛 Bug Fix
- [ ] 🔧 Optimization
- [x] 📝 Documentation
- [ ] ✅ Test
- [ ] 🐳 Other

## Related issues 

- Closes #990

## Checklist

- [x] Code follows style guide (ruff)
- [x] Tests added 
- [x] Documented the changes -> Documented through descriptive tests and this PR, please let us know if documentation in another location is needed.

## If you have comments or can explain your changes, please do so below.
To close #990 we now make sure that Pandas-like Dataframes can be indexed with tuples to support Polars-like behaviour.

### CODE REVIEW

File: tests/frame/getitem_dunder_test.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+class TestGetItemWorksForTupleIndexing:
+    """Tests added to address issue #990 - https://github.com/narwhals-dev/narwhals/issues/990"""
+
+    @pytest.fixture
+    def data(self) -> dict[str, list[int]]:
+        return {"x": [1, 2, 3]}
+
+    @pytest.fixture
+    def pd_df(self, data: dict[str, list[int]]) -> nw.DataFrame[Any]:
+        # TODO(mikeweltevrede): Use constructor with eager
```
Comment (mikeweltevrede) [inline tests/frame/getitem_dunder_test.py:20]: Still to do: use the constructor with eager mode

Comment (Riik) [reply tests/frame/getitem_dunder_test.py]: Done!

File: tests/frame/getitem_dunder_test.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+            ((0, 2), (0,)),
+        ],
+    )
+    def test_get_item_works_with_tuple_and_list_indexing(
+        self,
+        pd_df: nw.DataFrame[Any],
+        row_idx: list[int] | tuple[int],
+        col_idx: list[int] | tuple[int],
+    ) -> None:
+        pd_df = nw.from_native(pd_df, eager_only=True)
```
[the remark this answers is not in the document]
Comment (Riik) [reply tests/frame/getitem_dunder_test.py]: Fixed!

**Comment by windiana42** [comment]
This PR is ready for review.

File: tests/frame/getitem_test.py [hunk 3]
```diff
@@ -193,3 +195,58 @@ def test_slice_edge_cases(constructor_eager: Any) -> None:
...
+        ([0, 2], (0,)),
+        ((0, 2), (0,)),
+        ([0, 2], range(1)),
+        (range(2), [0]),
+        (range(2), range(1)),
+    ],
+)
+def test_get_item_works_with_tuple_and_list_indexing(
+    constructor_eager: Any,
+    row_idx: list[int] | tuple[int] | range,
```
Comment (mikeweltevrede) [inline tests/frame/getitem_test.py:214]: @Riik @windiana42 We can also type hint this as a `Sequence[int]`, correct?

File: .gitignore [hunk 4]
```diff
@@ -34,6 +34,3 @@ tpch/data/*
 
 # MacOS
 .DS_Store
-
-# IntelliJ IDEA
-.idea/
```
Comment (mikeweltevrede) [inline .gitignore]: @Riik why was this removed? I would consider this the same as `.vscode` but then for IntelliJ and Pycharm. It should not be pushed.

**Comment by mikeweltevrede** [comment]
@MarcoGorelli @Riik @windiana42 PS: Here is the video that showed me that `typing` is really just aliasing `collections` for a lot of type hints: https://www.youtube.com/watch?v=cv1F_c66utw

File: .gitignore (hunk 4 above)
Comment (Riik) [reply .gitignore]: I agree it would be easier to have .idea in gitginore as well, but figured it was easier to keep that separate from this PR since it was just about the bug ticket

File: tests/frame/getitem_test.py (hunk 3 above)
Comment (Riik) [reply tests/frame/getitem_test.py:214]: Yeah I think that is cleaner! Since 'Sequence' is what we were trying to support

File: .gitignore (hunk 4 above)
Comment (mikeweltevrede) [reply .gitignore]: I will remove the README clean up then too :)

File: tests/frame/getitem_test.py (hunk 3 above)
Comment (mikeweltevrede) [reply tests/frame/getitem_test.py:214]: On second thought, I am not sure we should because `Sequence` includes strings. We would want a `SequenceNotStr` type hint, which I don't think we want to introduce now.

**Review by MarcoGorelli — APPROVED** [review]
thanks @Riik @windiana42 @mikeweltevrede !
